### PR TITLE
enable request-body parameters for content-type `application/x-www-form-urlencoded` POSTs for long queries - RIAK-1875

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
Add ability to handle POSTs for search queries, when given the content-type **application/x-www-form-urlencoded**, or return a `415` if given another CT on POST. Should solve this issue: https://github.com/basho/yokozuna/issues/503, which references  RIAK-1875. 

Run riak_test w/ `riak/2.0` branch -> https://github.com/basho/riak_test/tree/riak/2.0/tests.

@kuenishi @shino @JeetKunDoug @fadushin may want to take a gander/review.

### CODE REVIEW

File: riak_test/yz_rt.erl [hunk 1]
```diff
@@ -58,6 +60,22 @@ create_index(Node, Index, SchemaName, NVal) ->
                [Index, SchemaName, NVal, Node]),
     ok = rpc:call(Node, yz_index, create, [Index, SchemaName, NVal]).
 
+-spec create_index_http(nodes(), index_name()) -> ok.
```
Comment (shino) [inline riak_test/yz_rt.erl]: `compile-riak-test` fails here:


```
Compiling riak_test/yz_rt.erl failed:
riak_test/yz_rt.erl:69: type nodes() undefined
ERROR: riak_test_compile failed while processing /home/shino/b/g/yokozuna-post-qeury: rebar_abort
```

File: riak_test/yz_rt.erl [hunk 2]
```diff
@@ -58,6 +60,22 @@ create_index(Node, Index, SchemaName, NVal) ->
...
+create_index_http(Cluster, Index) ->
+    Node = yz_rt:select_random(Cluster),
+    HP = hd(host_entries(rt:connection_info([Node]))),
+    create_index_http(Cluster, HP, Index).
+
+-spec create_index_http(nodes(), {string(), portnum()}, index_name()) -> ok.
```
Comment (shino) [inline riak_test/yz_rt.erl]: `compile-riak-test` fails here:

```
Compiling riak_test/yz_rt.erl failed:
riak_test/yz_rt.erl:69: type nodes() undefined
ERROR: riak_test_compile failed while processing /home/shino/b/g/yokozuna-post-qeury: rebar_abort
```

File: src/yz_wm_search.erl [hunk 3]
```diff
@@ -107,24 +105,28 @@ forbidden(RD, Ctx) ->
...
+        _ ->
+            {{halt, 415}, RD, Ctx}
     end.
 
 search(Req, S) ->
+    Params = wrq:req_qs(Req),
+    search(Req, S, Params).
+
+search(Req, S, []) ->
+    search(Req, S);
```
Comment (shino) [inline src/yz_wm_search.erl]: I guess this recursion to arity 2 function is for empty body in POST
search request to fallback to search by qeury string.  When GET search
request without query string (in HTTP sense) or POST one with empty
body and empty query string, then infinite recursion between arity 2
and 3.  Such requests are useless, but I don't know users won't do
such things.

Sample requests:

```
## GET
% curl -s 'http://127.0.0.1:8098/search/query/people'

## POST
% curl -v -X POST -H 'Content-Type: application/x-www-form-urlencoded' \
     -s 'http://127.0.0.1:8098/search/query/people'
```

**Comment by shino** [comment]
Thanks for nice fix for large query string, @zeeshanlakhani 
I agree with POST body approach :+1: 

Sorry for stepping-in while @fadushin has already assigned himself,
but as my self-training for yz/solr :sweat_smile: , I read diff and did some tests.
Confirmed query in POST body works well :confetti_ball: 
In the way, I found one tiny build error and one (hopefully harmless) infinite loop possibility,
so added comments for them in diff view.

File: riak_test/yz_rt.erl (hunk 2 above)
Comment (zeeshanlakhani) [inline riak_test/yz_rt.erl]: Hmm, not sure how I got this to build haha. I'll replace w/ `cluster()`, which is what we have already.

**Comment by shino** [comment]
Code diff looks nice and all riak_test under this repository succeeded. (Thanks for helping solving env errors)
Big :+1: from me :dizzy:

**Comment by shino** [comment]
Test results by @ksauzz , for 2.1 branch with the change of this PR. POST works nice for large query string.

**Comment by zeeshanlakhani** [comment]
@borshop merge

**Pull request merged**